privacy-protections/fingerprinting/main.js

```diff
@@ -0,0 +1,230 @@
+                if (idx < 2) {
+                    return;
+                }
+
+                path += '→ ' + segment;
+            });
+
+            const testA = resultsAObj[testId];
+            const testB = resultsBObj[testId];
+
+            if (change.kind === 'E') { // modified property
+                li.innerHTML = `<strong>${testId}</strong> ${path} - value <span class='changed'>changed</span> from <code>${JSON.stringify(change.lhs, null, 2)}</code> to <code>${JSON.stringify(change.rhs, null, 2)}</code>`;
+            } else if (change.kind === 'A') { // array change
+                if (change.item.kind === 'N') {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - new item <code>${JSON.stringify(change.item.rhs, null, 2)}</code> <span class='added'>added</span>`;
+                } else if (change.item.kind === 'D') {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - item <code>${JSON.stringify(change.item.lhs, null, 2)}</code> <span class='removed'>removed</span>`;
+                }
+            } else if (change.kind === 'D') { // deleted property
+                if (change.item) {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - item <code>${JSON.stringify(change.item.rhs, null, 2)}</code> <span class='removed'>removed</span>`;
+                } else {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - property <span class='removed'>missing</span>`;
+                }
+            } else if (change.kind === 'N') { // added property
+                if (change.item) {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - new item <span class='added'>added</span> <code>${JSON.stringify(change.item.lhs, null, 2)}</code> (<code>${JSON.stringify(testB.value, null, 2)}</code>)`;
+                } else {
+                    li.innerHTML = `<strong>${testId}</strong> ${path} - <span class='added'>new</span> property (<code>${JSON.stringify(change.rhs, null, 2)}</code>)`;
+                }
+            }
+
+            list.appendChild(li);
+        });
+    }
+
+    diffDetailsDiv.innerHTML = '';
+    diffDetailsDiv.appendChild(list);
+
+    diffDiv.removeAttribute('hidden');
+}
+
+saveToLS.addEventListener('click', () => {
+    localStorage.setItem('results', JSON.stringify(results, null, 2));
+});
+
+compareWithLS.addEventListener('click', () => {
+    const oldResults = JSON.parse(localStorage.getItem('results'));
+
+    compareResults(oldResults, results);
+});
+
+compareWithFile.addEventListener('change', event => {
+    const reader = new FileReader();
+    reader.onload = () => {
+      const oldResults = JSON.parse(reader.result);
+
+      compareResults(oldResults, results);
+    };
+    reader.readAsText(compareWithFile.files[0]);
+})
+
+downloadButton.addEventListener('click', () => downloadTheResults());
+
+// run tests if button was clicked or…
+startButton.addEventListener('click', () => runTests());
```

Comment:
One thing I thought of is that by making this run on click, we won't be testing any race conditions around fingerprinting. (For instance, a site fingerprints with a script in head right away, without waiting for the dom to load or fully load). Depending on how defenses are implemented, they may or may not defend against that. Probably not a problem for our use case, but maybe something to be aware of.

Reply:
Good point, I haven't thought about this. However(!) there is a way to run test on page load - see code below. I added it mostly for automated tests, but as far as I understand it would also be useful in the scenario you are describing.